### ISSUE TITLE
Side effects in the new EDD_Payment class - Orders may be saved and loaded with wrong totals

### ISSUE THREAD

**Issue by daigo75** [easydigitaldownloads/Easy-Digital-Downloads#4253]
After spending days trying to trace a "mysterious" issue that appeared with EDD 2.5, I noticed that the new logic used to save orders' metadata has some side effects, that cannot be handled effectively by external plugins. One of those side effects is that the hooks related to the processing of orders meta are triggered multiple times, and this can cause issues.

In my specific case, the issue relates to the order amounts in a multi-currency environment. For reporting purposes, order amounts must be saved in two currencies:
* In original currency, as standard.
* In shop's base currency, for reporting purposes (reports make a "brutal" `SUM()` of order totals, therefore it's vital that those values are in the same currency).

For example, an order for 100 EUR will have two totals:
* 100 EUR (order currency)
* 108.94 USD (base currency, for reporting purposes).

To make this transparent to EDD, the orders are saved as follows:
* The original total (100 EUR) is moved from `_edd_payment_total` to a custom field.
* The value of `_edd_payment_total` is replaced by 108.94 USD.

As a result:
* When an order is loaded (view order, payment history, etc), there is a hook that "swaps" the totals dynamically, thus showing the correct amount (100 EUR).
* When a report is prepared, it calculates a sum of `_edd_payment_total` (this cannot be changed, as queries are hard-coded). This also works, and shows a report total in base currency, which is accurate.

This worked perfectly until EDD 2.5 introduced the new `EDD_Payment` class. The new class relies on its `save()` method to update the order meta. Such method saves the payment totals whether they have changed or not, and triggers multiple conversions of the same amounts.

How the logic worked until EDD 2.4
--
1. Total: 100 EUR
2. Save order total -> Call `_edd_payment_total` hook.
3. A filter performed the conversion (100 EUR = 108.94 USD) and saved the two values.
4. When the payment was set to "completed", only the `status` meta was saved. The filter from 3 was no longer invoked.

How the logic works in EDD 2.5+
--
1. Total: 100 EUR
2. Save order total -> Call `_edd_payment_total` hook.
3. A filter performs the conversion (100 EUR = 108.94 USD) and saves the two values.
4. The status changes to "completed". This involves loading the order that was just saved, and that fetches the order total that was converted earlier, i.e. 108.94. The `EDD_Payment::save()` method is then called, and it saves the order total again, passing the converted value. The conversion hook takes it as face value, and converts it again, returning 118.71.

This repeated conversion will occur every time the order total is saved. Order is loaded, total is 118.71. `EDD_Payment::save()` will save it again, triggering the conversion and changing the total to 129.35. 

Currently, there is no way to determine if the conversion should be performed or not, therefore the unnecessary conversions cannot be prevented. Adding some logic to perform the conversion only once doesn't work either, as a payment could be trying to save the same original amount twice (it happens with PayPal Standard, when the IPN is processed) and **both** calls should trigger a conversion.

How to fix the issue (maybe)
--
The `EDD_Payment` class lacks events or filters to determine the context of its calls, thus making it practically impossible to determine what it's doing. Adding some actions that can help following its flow should make it easier to check what the class does and when, and control the outcome of the operations.
I'm going to experiment with it, to see if I can find a possible solution.

**Comment by daigo75** [comment]
Pre-emptive answer, in case anyone wonders why bothering with this whole "swap" business: that was the only way to make reports work. Reports are quite rigid, they run hard-coded queries against set fields. Since this could not be changed, I had to find a way to have those "core" fields contain consistent data. The solution was to store base currency data in them, and move the "mixed currency" data elsewhere. 

For reference, and possible future improvement, this is the opposite of the logic I originally implemented in WooCommerce. In WC, reports have filters that allow to tell them which metadata they should load. The only swap occurs when a report is run: core fields contain order data in order currency (no swaps), and the base currency data is stored separately, invisible to WC itself. When a report runs, a filter tells it to ignore the `_order_total` field and take `_order_total_base_currency` instead. Other filters also allow to change the `WHERE` clause, so that it's possible, for example, to only load sales in a specific currency, and have multiple reports showing the sub-totals separately, in each currency.

Having such type of filters in EDD would be greatly beneficial, as it would make reports more flexible.

**Comment by pippinsplugins** [comment]
Excellent catch.

I like your idea of adding a `edd_pre_setup_payment` hook.

Could you show me how you would use that hook inside of your plugin to resolve the issue, just for demonstration purposes?

**Comment by daigo75** [comment]
Yes, the `pre_` hook would make it easier to load a payment correctly without having to "adjust" its properties after the class has been fully initialised. It will also allow to decouple filters from the actual class implementation, as they will alter the data retrieved from the database, and the class itself will simply pick the elements it needs and stored them into its own properties.

Making a generic example is difficult, I spotted the issue due to the specific way I handle the multi-currency amounts. Here's an example of how the solution (which I already tested) works.

1. A payment is created for 100 EUR + 20 EUR tax. There is one item, whose value is, thus, 100 EUR + 20 EUR tax.
2. The payment is saved to the database. The class will save the following:
  * order currency = EUR
 * `_edd_payment_total` = 100
  * `_edd_payment_tax` = 20
  * `_edd_payment_meta` (array)
     - Item total = 100
     - Item tax = 20

From the `EDD_Payment` class' perspective, that's all that happened. The database contains 100 EUR + 20 EUR tax. The object's properties are not changed (i.e. it's stil an instance with 100 EUR + 20 EUR).
3. When the meta from _2_ is saved, several, the following filters alter it:
  *  `edd_update_payment_meta__edd_payment_total`
  * `edd_update_payment_meta__edd_payment_tax`
  * `edd_update_payment_meta__edd_payment_meta`

Specifically, the original total (100 EUR) is saved in `_edd_payment_total_order_currency` and `_edd_payment_total` is replaced by its USD equivalent. The same applies to `_edd_payment_tax` and to the price attributes in `_edd_payment_meta` (which are stored in separate fields, within the same array).
4. When the payment is loaded again from the database, it reads the fields it originally saved (see point 2). The `EDD_Payment` class doesn't know, and should not know, that someone played a "three card trick" and swapped the values.
The `edd_pre_setup_payment` will fire before that happens, and it will allow to set "reverse filters" to restore the original values. The filters, in my case, would be set as follows:

~~~
add_action('edd_pre_setup_payment', array($this, 'load_order_in_order_currency'), 5);
// [...]

/**
 * Adds filters that ensures that order totals are loaded and returned in
 * order currency, rather than base currency.
 */
public function load_order_in_order_currency() {
  // Replace the value of _edd_payment_total with _edd_payment_total_order_currency
  add_filter('edd_get_payment_meta__edd_payment_total', array($this, 'edd_get_payment_total_in_order_currency'), 1, 2);
  // Replace the value of _edd_payment_tax with _edd_payment_tax_order_currency
  add_filter('edd_get_payment_meta__edd_payment_tax', array($this, 'edd_get_payment_tax_in_order_currency'), 1, 2);
  // Replace the price data in _payment_meta with their order currency value
  add_filter('edd_get_payment_meta', array($this, 'edd_get_payment_meta_in_order_currency'), 10, 3);
~~~

Flow summary (pseudo-code)
--
1. `EDD_Payment->save(_edd_payment_total = 100 EUR, _edd_payment_tax = 20 EUR)`
2. On save, filters perform the following replacements:
  * `_edd_payment_total_order_currency `= 100 EUR
  * `_edd_payment_total `= 108.94 USD
  * `_edd_payment_tax_order_currency `= 20 EUR
  * `_edd_payment_tax `= 21.88 USD
And so on. Data is then saved to the database.

3. `EDD_Payment->load()`
4. On `edd_pre_setup_payment` event, set "on load" filters.
5. On load, filters replace the requested values on the fly:
  * `_edd_payment_total` with `_edd_payment_total_order_currency` = 100 EUR
  * `_edd_payment_tax` with `_edd_payment_tax_order_currency` = 20 EUR
Etc. The instance of `EDD_Payment` loaded from the database will, therefore, contain the same values that it had when it was saved.

**Comment by pippinsplugins** [comment]
Perfect. I like it.

**Comment by pippinsplugins** [comment]
@cklosowski Any objections?

**Comment by cklosowski** [comment]
I don't think so. It's clearly not breaking any unit tests.

I think, if possible, I'd like to see a unit test that actually tests this, and verifies it's working as expected if that's possible @daigo75. Just so have a bit of an automated way to test that we don't break it in any future implementations.

**Comment by pippinsplugins** [comment]
I'd like to go ahead and get it merged in if you're not opposed.

**Comment by cklosowski** [comment]
I think I'm good with it, let's merge and get some further testing with `EDD_Payment`, especially with the refund work in #4223 that's coming.

**Comment by cklosowski** [comment]
Going to merge and keep the issue open for verification.

**Comment by pippinsplugins** [comment]
@daigo75 Are you able to test one last time with the PR merged in?

**Comment by daigo75** [comment]
It seems to work fine, from what I've seen. :+1: 

I'm also going to submit another PR, for a different issue, but still on the same topic.

**Comment by pippinsplugins** [comment]
Glad to hear it!
